File: python/test1.py
import time
import psutil

INPUT_CSV_FILE = "./data/sales.csv"
OUTPUT_CSV_FILE = "./python/salesWithTaxes.csv"
BENCHMARK_CSV_FILE = "./output/benchmarks.csv"
TAX_PERCENT = 0.05
log_bench = False


def benchmark(tag):
    time_start = time.perf_counter()
    memory_start = psutil.virtual_memory().used

    def finish_benchmark():
        # must be taken to miliseconds
        time_usage = f"{round((time.perf_counter() - time_start) * 1000, 2)} ms"
        memory_usage = f"{round((psutil.virtual_memory().used - memory_start)/1024/1024, 2)} mb"
        if log_bench:
            print("-------- BENCHMARK", tag, "RESULTS --------")
            print("time:", time_usage, "memory:", memory_usage)
            print("-------- END BENCHMARK RESULTS --------\n")
        with open(BENCHMARK_CSV_FILE, 'a') as f:
            f.write(f"python,{tag},{time_usage},{memory_usage}\n")
    return finish_benchmark


def read_file():
    read_bench = benchmark("read file")
    with open(INPUT_CSV_FILE) as file:
        content = file.readlines()
    header = content[:1]
    sales = []
    for line in content[1:]:
        sales.append(line.split(","))
    read_bench()
    process_data(sales, header)


def process_data(sales, header):
    process_bench = benchmark("process data")
    sales_with_taxes = []
    for sale in sales:
        price = float(sale[2][1:])
        priceWithTax = round(price + (price * TAX_PERCENT), 2)
        sales_with_taxes.append([sale[0], sale[1], f"${priceWithTax}"])
    lines = [",".join(header)]
    for sale in sales_with_taxes:
        lines.append(",".join(sale)+"\n")
    write_data = "".join(lines)
    write_file(write_data)
    process_bench()


def write_file(write_data):
    write_bench = benchmark("write file")
    with open(OUTPUT_CSV_FILE, "w") as output:
        output.write(write_data)
    write_bench()


def main():
    read_file()


if __name__ == "__main__":
    main()

File: js/test1.js
const fs = require('fs/promises')

const INPUT_CSV_FILE = './data/sales.csv'
const OUTPUT_CSV_FILE = './js/salesWithTaxes.csv'
const BENCHMARK_CSV_FILE = './output/benchmarks.csv'

const logBench = false
const TAX_PERCENT = 0.05

function benchmark(tag) {
  const startTime = performance.now()
  const memoryStart = process.memoryUsage().heapUsed
  return () => {
    const totalTime = (performance.now() - startTime).toFixed(2) + ' ms'
    const memoryUsage =
      ((process.memoryUsage().heapUsed - memoryStart) / 1024 / 1024).toFixed(
        2
      ) + ' mb'

    if (logBench) {
      console.log(`------ BENCHMARK ${tag.toUpperCase()} RESULTS ------`)
      console.log(`| Time: ${totalTime}, Memory: ${memoryUsage} |`)
      console.log('------ END BENCHMARK RESULTS ------\n')
    }

    // Log to file
    fs.appendFile(
      BENCHMARK_CSV_FILE,
      `javascript,${tag},${totalTime},${memoryUsage}\n`
    ).catch((error) => {
      throw error
    })
  }
}

function readFile() {
  const benchInput = benchmark('read file')
  const sales = []

  fs.readFile(INPUT_CSV_FILE, 'utf8')
    .then((data) => {
      data.split('\r\n').forEach((line) => {
        sales.push(line.split(','))
      })
      processData(sales)
      benchInput()
    })
    .catch((error) => {
      throw error
    })
}

function processData(sales) {
  const benchProcess = benchmark('process data')
  const salesWithTaxes = sales.map((data, i) => {
    if (i > 0 && data.length === 3) {
      const price = Number(data[2].slice(1))
      const priceWithTax = (price + price * TAX_PERCENT).toFixed(2)
      data[2] = `$${priceWithTax}`
    }
    return data
  })
  const writeData = salesWithTaxes.reduce(
    (prev, curr) => `${prev}${curr.join(',')}\n`,
    ''
  )
  writeFile(writeData)
  benchProcess()
}

function writeFile(writeData) {
  const benchOutput = benchmark('write file')

  fs.writeFile(OUTPUT_CSV_FILE, writeData, 'utf8').catch((error) => {
    throw error
  })
  benchOutput()
}

function main() {
  readFile()
}
main()
